### ISSUE TITLE
Unable to register global express middleware from appium plugin 

### ISSUE THREAD

**Issue by sudharsan-selvaraj** [appium/appium#17411]
## The problem

I'm currently building an appium plugin and I need to add an express middleware to handle all incoming API requests before it is handled by the appium server. When i tried adding the middleware inside `updateSever` methods as below,
```
public static async updateServer(expressApp: any, httpServer: any, cliArgs: any): Promise<void> {
  expressApp.use("/", (requuest, repsonse, next)=>{
    console.log(request.url)
  })
}
```
The middleware is not being called because appium routes are registered before the middleware and all requests are severd automatically by the matching appium route.

One hacky way to solve this issue is to manually update the express route stack and inject the middleware before the appium routes are registered

```
public static async updateServer(expressApp: any, httpServer: any, cliArgs: any): Promise<void> {
  let indexToPushMiddleware = expressApp._router.stack.findIndex((s: any) => s.route);
  expressApp.use("/", (requuest, repsonse, next)=>{
    console.log(request.url)
  });
 expressApp._router.stack.splice(indexToPushMiddleware, 0, expressApp._router.stack.pop());
}
```

Sample output of `console.log(expressApp._router.stack)` is https://gist.github.com/sudharsan-selvaraj/50729d447356dcb6563bbf8652ea6a94

**Comment by saikrishna321** [comment]
@jlipps @boneskull What do you think about think? We are also doing the same in appium-device-farm for remote proxy the command. If anything from `expressApp` changes we need to change this hack as well.

**Comment by jlipps** [comment]
yeah. i'm not sure we can cover all use cases with the express app in a clean way. at the moment i think using the "hack" is probably the right way. unless @boneskull has any ideas.

in the future we could add first class support for adding middleware to the express server, seems like it would need to be complex because there might be various places an extension could want the middleware to live.

**Comment by boneskull** [comment]
yeah other than this workaround, that's what we can do right now. to support this exact use case, we'd need to split the server initialization into an "instantiate express" and "create routes" portions, then run `updateServer` between the two... but you'd also want to run `updateServer` _after_ Appium creates its own routes so that they can be modified.  I'm not sure there's anything to be done there, as your use-case is probably the more unusual of the two.  

another way to approach this would be some sort of `preServer` method might be useful to monkey w/ options provided to the server factory function.  you could then replace the `routeConfiguringFunction` option (which creates all the routes) and replace it with smth like `_.flow(myRouteConfiguringFunction, routeConfiguringFunction)`.

otherwise, if you want a nicer way to prepend routes, you may want to create an issue on express's repo.  good luck 😉

**Comment by boneskull** [comment]
fwiw, I think that events make a lot of sense for some of this stuff.  appium would emit events for "created server config", "created server", "created routes", "server listening" and a plugin could subscribe to them.  the exception-related drawbacks of `EventEmitter` can be worked around via a nicer 3p EE replacement package.

**Comment by mykola-mokhnach** [comment]
@sudharsan-selvaraj Would the feature https://github.com/appium/appium/pull/22632 work for what you are trying to achieve?

**Comment by mykola-mokhnach** [comment]
Closed as fixed
